consensus/storage_fault_monitor.go

```diff
@@ -0,0 +1,89 @@
+package consensus
+
+import (
+	"context"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+
+	"github.com/filecoin-project/go-filecoin/abi"
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/filecoin-project/go-filecoin/vm/errors"
+)
+
+// monitorPlumbing is an interface for the functionality StorageFaultSlasher needs
+type monitorPlumbing interface {
+	MessageQuery(ctx context.Context, optFrom, to address.Address, method string, params ...interface{}) ([][]byte, error)
+}
+
+// slashingMsgOutbox is the interface for the functionality of Outbox StorageFaultSlasher needs
+type slashingMsgOutbox interface {
```

Comment:
Why bother to add message to outbox if we're not going to broadcast it? Wouldn't it be just as easy to add it to the inbox? Isn't that all adding all the outbox does aside from broadcasting?

Reply:
Outbox Send does a bunch of other checks as well as determining current height and nonce, validates the message,  then it Enqueues the message into the MessageQueue, and _then_ calls Publish which is strictly not true; it not only adds to mpool but also itself will broadcast or not.  Inbox's `Add` adds directly to the message pool and doesn't add to the MessageQueue.  It seems like the slashing message should be following the same code path as every other message sent to actors.

Reply:
Non-Blocking:
Ok, looking a little closer at it, I sort of get what your saying. I know this was mostly covered in the setup PR that introduced the bcast flag, but I didn't get a chance to comment on that. Needing to add flags like this always make me question our abstractions

Our goal is to create a message and add it directly to our message pool. The weird thing here for me is that we do this through our `Outbox` interface. It's even weirder that Outbox calls `Publish` on the message publisher with a flag that tells it not to publish, just so we can add it to the message pool. 

`Outbox.Send` is responsible for:
* Message construction from components
* Message signing
* Nonce generation
* Retrieving the block height
* "Publishing" through the `Publisher` interface

`defaultMessagePublisher.Publish` is responsible for:
* Encoding the message for publishing
* Adding the message to the message pool ensuring that miner's get their own messages. I think this is due to a bug in libp2p.
* Possibly publishing the encoded message to the network

In my ideal world, there would be a new `MessageSigner` struct that would export a `NewSignedMessage` method to construct, sign and nonce a message. Also, message pool would get its access to the block height back, since the need for it is an implementation detail that doesn't really need to be shared with Outbox or defaultMessagePublisher. It could be this goes into something like `Policy` and that is what satisfies something like a `AddToMessagePool` interface.

Then `Outbox.Send` would:
* Construct the signed, nonce'd message with a call to `MessageSigner`.
* Add the message to the message pool (assuming this is still necessary)
* Publish the message through the message publisher.

`defaultMessagePublisher.Publish` would just encode and publish the message to the network as its name implies.

Then code like `StorageFaultMonitor` would could add messages to the local message pool with dependencies on `NewSignedMessage` and `AddToMessagePool`.

CC @anorth

Reply:
ok I think this means I'm to consider these changes in another PR @acruikshank ?

Reply:
Ack. See also #3052 - the pool and queue may become the same thing.